File: paintselector/src/main/java/com/sasaj/graphics/paintselector/SaturationBrightnessSelector.java
package com.sasaj.graphics.paintselector;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ComposeShader;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.Shader;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;

import com.sasaj.graphics.paintselector.com.sasaj.graphics.paintselector.interfaces.PaintSelector;

/**
 * Created by User on 7/17/2016.
 */
public class SaturationBrightnessSelector extends View{

    private static final String TAG = "SaturationBrightness";
    private int sourceWidth = 0;
    private int sourceHeight = 0;

    public float hue;
    public float saturation;
    public float brightness;
    public PaintSelector picker;
    private float cx;
    private float cy;
    private Paint backgroundPaint;
    private Paint circlePaint;
    private float factor;

    final float[] hsv = { 1f, 1f, 1f };
    private Shader brightnessGradient;
    private Shader saturationGradient;

    public SaturationBrightnessSelector(Context context) {
        super(context);
        init();
    }

    public SaturationBrightnessSelector(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public SaturationBrightnessSelector(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {

        setLayerType(View.LAYER_TYPE_SOFTWARE, null);
        factor = 0;

        circlePaint = new Paint();
        circlePaint.setStrokeWidth(5);
        circlePaint.setAntiAlias(true);
        circlePaint.setStyle(Paint.Style.STROKE);
        circlePaint.setStrokeJoin(Paint.Join.ROUND);
        circlePaint.setStrokeCap(Paint.Cap.ROUND);
    }

    public void setColorPicker(PaintSelector picker){
        this.picker = picker;

        int color = picker.getColor();
        float[] hsv = getHsvFromColor(color);
        hue = hsv[0];
        saturation = hsv[1];
        brightness = hsv[2];
    }


    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        sourceWidth = w;
        sourceHeight = h;
        factor = 1/(float)w;
        cx = saturation/factor;
        cy = (1 - brightness)/factor;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int width = MeasureSpec.getSize(widthMeasureSpec);
        setMeasuredDimension(width, width);
    }


    public void setHue(float hue){
        this.hue = hue;

        hsv[0] = hue;
        picker.setColor(getColorFromHsv());

        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (backgroundPaint == null) {
            backgroundPaint = new Paint();
            brightnessGradient = new LinearGradient(0f, 0f, 0f, sourceHeight, 0xffffffff, 0xff000000, Shader.TileMode.CLAMP);
        }
        int rgb = Color.HSVToColor(hsv);
        saturationGradient = new LinearGradient(0f, 0f, sourceWidth, 0f, 0xffffffff, rgb, Shader.TileMode.CLAMP);
        ComposeShader shader = new ComposeShader(brightnessGradient, saturationGradient, PorterDuff.Mode.MULTIPLY);
        backgroundPaint.setShader(shader);
        canvas.drawRect(0f, 0f, this.getMeasuredWidth(), this.getMeasuredHeight(), backgroundPaint);

        float circleHue = hue + 180; if(circleHue > 360) circleHue = circleHue - 360;
        int circleColor = Color.HSVToColor(new float []{circleHue, 1- cx * factor, cy*factor});
        circlePaint.setColor(circleColor);
        if(cx<0) cx=0; if(cx > sourceWidth) cx = sourceWidth; if (cy < 0) cy = 0; if(cy > sourceHeight) cy = sourceHeight;
        canvas.drawCircle(cx, cy, 10, circlePaint);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        cx = event.getX();
        cy = event.getY();
        saturation = cx*factor;
        brightness = 1 - cy*factor;

        int color = getColorFromHsv();
        picker.setColor(color);
        invalidate();
        return true;
    }

    private float[] getHsvFromColor(int color) {
        float[] hsv = new float[3];
        Color.colorToHSV(color, hsv);
        return hsv;
    }

    private int getColorFromHsv(){
        float[] hsv;
        hsv = new float[]{hue, saturation, brightness};
        return Color.HSVToColor(hsv);
    }
}

File: app/src/main/java/com/sasaj/graphics/drawingapp/ui/brushselector/views/SaturationBrightnessSelector.kt
package com.sasaj.graphics.drawingapp.ui.brushselector.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.ComposeShader
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.Shader
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

import com.sasaj.graphics.drawingapp.ui.brushselector.interfaces.PaintSelector


/**
 * Created by sjugurdzija on 7/17/2016.
 */
class SaturationBrightnessSelector : View {
    private var sourceWidth = 0
    private var sourceHeight = 0

    var selectedHue: Float = 0.0f
    var saturation: Float = 0.0f
    var brightness: Float =0.0f
    lateinit var picker: PaintSelector
    private var cx: Float = 0.0f
    private var cy: Float = 0.0f
    private lateinit var backgroundPaint: Paint
    private lateinit var circlePaint: Paint
    private var factor: Float = 0.0f

    private val globalHsv = floatArrayOf(1f, 1f, 1f)
    private var brightnessGradient: Shader? = null
    private var saturationGradient: Shader? = null

    private val colorFromHsv: Int
        get() {
            val hsv: FloatArray = floatArrayOf(selectedHue, saturation, brightness)
            return Color.HSVToColor(hsv)
        }

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {

        setLayerType(View.LAYER_TYPE_SOFTWARE, null)
        factor = 0f

        backgroundPaint = Paint()

        circlePaint = Paint()
        circlePaint.strokeWidth = 5f
        circlePaint.isAntiAlias = true
        circlePaint.style = Paint.Style.STROKE
        circlePaint.strokeJoin = Paint.Join.ROUND
        circlePaint.strokeCap = Paint.Cap.ROUND
    }

    fun setColorPicker(picker: PaintSelector) {
        this.picker = picker

        val color = picker.color
        val hsv = getHsvFromColor(color)
        selectedHue = hsv[0]
        saturation = hsv[1]
        brightness = hsv[2]
    }


    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        sourceWidth = w
        sourceHeight = h
        factor = 1 / w.toFloat()
        cx = saturation / factor
        cy = (1 - brightness) / factor

        updateView()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val width = View.MeasureSpec.getSize(widthMeasureSpec)
        setMeasuredDimension(width, width)
    }


    fun setHue(hue: Float) {
        this.selectedHue = hue

        globalHsv[0] = hue
        picker.color = colorFromHsv

        updateView()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        canvas.drawRect(0f, 0f, this.measuredWidth.toFloat(), this.measuredHeight.toFloat(), backgroundPaint)
        canvas.drawCircle(cx, cy, 10f, circlePaint)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        cx = event.x
        cy = event.y
        saturation = cx * factor
        brightness = 1 - cy * factor

        val color = colorFromHsv
        picker.color = color

        updateView()

        return true
    }

    private fun getHsvFromColor(color: Int): FloatArray {
        val hsv = FloatArray(3)
        Color.colorToHSV(color, hsv)
        return hsv
    }

    private fun updateView() {

        val rgb = Color.HSVToColor(globalHsv)

        //Update view's background parameters
        brightnessGradient = LinearGradient(0f, 0f, 0f, sourceHeight.toFloat(), -0x1, -0x1000000, Shader.TileMode.CLAMP)
        saturationGradient = LinearGradient(0f, 0f, sourceWidth.toFloat(), 0f, -0x1, rgb, Shader.TileMode.CLAMP)
        val shader = ComposeShader(brightnessGradient!!, saturationGradient!!, PorterDuff.Mode.MULTIPLY)
        backgroundPaint.shader = shader

        //Update indicator's parameters
        var circleHue = selectedHue + 180
        if (circleHue > 360) circleHue -= 360
        val circleColor = Color.HSVToColor(floatArrayOf(circleHue, 1 - cx * factor, cy * factor))
        circlePaint.color = circleColor
        if (cx < 0) cx = 0f
        if (cx > sourceWidth) cx = sourceWidth.toFloat()
        if (cy < 0) cy = 0f
        if (cy > sourceHeight) cy = sourceHeight.toFloat()

        invalidate()
    }

    companion object {

        private val TAG = "SaturationBrightness"
    }
}

















